Feature/andrzejs text file gen

therealendrju opened the pull request:
Add Offset to TextFileGen module This allows controlling it via a script in grind jobs, making running grid jobs easier. 
New parameter defaults to zero, should be invisible to users who don't care about it. 
Restructured the code - moved the event reading into a member function.

Commit:
move out file reading into a helper function to simplify next steps

Commit:
add Offset parameter, allowingto skip n events in a file

Commit:
TextFileGen final clean up

Commit:
Merge branch 'develop' of github.com:LArSoft/larsim into feature/andrzejs_TextFileGen

Commit:
add the extra parameter to the .fcl file for reference

FNALbuild commented:
A new Pull Request was created by  @therealendrju (Andrzej Szelc) for develop.

It involves the following packages:

larsim


@LArSoft/level-1-managers, @LArSoft/level-2-managers can you please review it and eventually sign? Thanks.

cms-bot commands are listed <a href="https://cdcvs.fnal.gov/redmine/projects/larsoft/wiki/Pull_request_comments_that_trigger_CI_actions">here</a>

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//1100/console

knoepfel commented:
Thanks @therealendrju for the PR.  It's clear what your intention is--however, I think a cleaner implementation can be achieved without having to make as many changes as you did.  The important aspect is to skip the first *n* events--and that's something that can be done in `beginJob` with the following code:

```c++
void beginJob() 
{
  fInputFile = new std::ifstream(fInputFileName.c_str());
  if ( *fInputFile )
     throw cet::exception("TextFileGen") << "input text file "
 					<< fInputFileName
 					<< " cannot be read.\n";

    std::istringstream buffer;
    for (unsigned i = 0; i != fOffset; ++i) {
        std::string line;
        getline(iss, line);
        buffer.str(line);

        // Parse read line for the event number and particles per event
        unsigned event, nparticles;
        buffer >> event >> nparticles;
        for (unsigned p = 0; p != nparticles; ++p) {
            constexpr auto all_chars_until = std::numeric_limits<unsigned>::max();
            iss.ignore(all_chars_until, '\n');
        }
    }
}
```

Then the `produce` function can be how it was before.

therealendrju commented:
Hi @knoepfel ,
thanks! I agree that moving the initial jumping to the offset to beginJob is definitely a cleaner solution.
But if I use your method, then we effectively have two pieces of code to do the same thing, i.e. read in the number of particles and skip the n-lines. I'd feel better if we had just one piece of code to do it. Can I propose to keep the external function and move its initial reading into the beginJob? Or is there a reason you don't want that code moved out to a function?
thanks!

knoepfel commented:
@therealendrju, breaking apart the code into more manageable bits (e.g. external functions) is a good thing--I am totally in favor of that!  But I think the granularity of those bits probably needs to be adjusted.  The main concern I have with your solution is that to skip to an offset of *n*, you have to create a vector *n* times, and do a lot of computation that you'll just throw away.  How about this...

1. Create utility to read event number and particles

```c++
std::pair<unsigned, unsigned> readEventInfo(std::istream& is)
{
  std::string line;
  getline(iss, line);
  std::istringstream buffer{line};

  // Parse read line for the event number and particles per event
  unsigned event, nparticles;
  buffer >> event >> nparticles;
  return {event, nparticles};
}
```

2. The `beginJob` function can then look like:
```c++
void beginJob() 
{
  fInputFile = new std::ifstream(fInputFileName.c_str());
  if ( *fInputFile )
     throw cet::exception("TextFileGen") << "input text file "
 					<< fInputFileName
 					<< " cannot be read.\n";

  for (unsigned i = 0; i != fOffset; ++i) {
    auto const [eventNo, nparticles] = readEventInfo(*fInputFile);
    for (unsigned p = 0; p != nparticles; ++p) {
       constexpr auto all_chars_until = std::numeric_limits<unsigned>::max();
       iss.ignore(all_chars_until, '\n');
    }
  }
}
```

3. You can still provide a `readNextHepEvt` function, but I would have it return a `simb::MCTruth` object:
```c++
simb::MCTruth readNextHepEvt() 
{
  // ...
  simb::MCTruth nextEvent;
  auto const [eventNo, nParticles] = readEventInfo(*fInputFile);
  for (unsigned short i = 0; i < nParticles; ++i) {
    ...
    simb::MCParticle part(i, pdg, "primary", firstMother, mass, status);
    ...
    nextEvent.Add(part);
  }
  return nextEvent;
}
```
The `produce` function is then pretty simple:
```c++
void evgen::TextFileGen::produce(art::Event & e)
{
  // check that the file is still good
  if( !fInputFile->good() )
    throw cet::exception("TextFileGen") << "input text file "
					<< fInputFileName
					<< " cannot be read in produce().\n";
   auto truthcol = std::make_unique<std::vector<simb::MCTruth>>();
   truthcol->push_back(readNextHepEvt());
   e.put(move(truthcol));
 }
```
What do you think?

therealendrju commented:
Hi @knoepfel ,
thanks again. That looks very elegant - I like it. I will try to implement that and get back to you.

cheers,
Andrzej

Commit:
modify code to not create vectors needlesly, move offset calculation to beginJob

Commit:
code cleanup

FNALbuild commented:
Pull request #85 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

therealendrju commented:
Hi @knoepfel, I have implemented your suggestions, 
and cleaned up after. Code builds and works as previously as far as I can tell.
thanks!
Andrzej

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//1101/console

knoepfel commented:
Looks good, @therealendrju.  I'll approve once there's L2 approval.

lgarren commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
+LArSoft tests OK on slf7 for c7:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/14155&builds= 
 +build

FNALbuild commented:
+LArSoft tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/14154&builds= 
 +build

FNALbuild commented:
-lariat failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/6230&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/14154&builds=

FNALbuild commented:
-uboone tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/6977&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/14154&builds=

FNALbuild commented:
+argoneut tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/6491&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/14154&builds=

FNALbuild commented:
-icarus tests warning, with build warning,, with ignored warning for build, on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:icarus/view_builds/index?offset=0&builds=icarus_ci/3472&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/14154&builds=

FNALbuild commented:
+sbnd tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:sbnd/view_builds/index?offset=0&builds=sbnd_ci/8427&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/14154&builds=

FNALbuild commented:
-dune tests warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/12588&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/14154&builds=

absolution1 commented on larsim/EventGenerator/TextFileGen_module.cc:
```diff
@@ -94,17 +97,19 @@ class evgen::TextFileGen : public art::EDProducer {
 //------------------------------------------------------------------------------
 evgen::TextFileGen::TextFileGen(fhicl::ParameterSet const & p)
   : EDProducer{p}
+  , fOffset{p.get<unsigned long int>("Offset",0)}
```
do you need the default value here?  Default values do make it harder to catch fcl mistakes...

therealendrju replied:
that makes it invisible to people who don't want this feature, i.e. people don't need to update their .fcl files if they don't want to.

absolution1 replied:
Is there a base FCL to update as that could absorb the change?  Otherwise I might argue that any new parameter is something an end user may want to know about.

knoepfel replied:
@therealendrju, are there any circumstances in which setting the offset value to something other than 0 is *necessary* to get the correct result?  If so, removing the C++ default of 0 (@absolution1's suggestion) may be the best option.

lgarren replied:
> @therealendrju, are there any circumstances in which setting the offset value to something other than 0 is _necessary_ to get the correct result? If so, removing the C++ default of 0 (@absolution1's suggestion) may be the best option.

@therealendrju - comments please?

therealendrju replied:
Hi, sorry. Yes, the value is what steers the behaviour of the module, so if you want to use the new functionality you need to set it to something different than zero. I'll remove the default value and uncomment the .fcl file to provide the zero there.

therealendrju replied:
@lgarren , @absolution1 , @knoepfel   - commit pushed.

Commit:
remove default Offset value (0) and uncomment it in the .fcl file

FNALbuild commented:
Pull request #85 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//1104/console

lgarren commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
+LArSoft tests OK on slf7 for c7:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/14213&builds= 
 +build

FNALbuild commented:
+LArSoft tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/14212&builds= 
 +build

FNALbuild commented:
-lariat failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/6255&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/14212&builds=

FNALbuild commented:
+argoneut tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/6517&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/14212&builds=

FNALbuild commented:
+sbnd tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:sbnd/view_builds/index?offset=0&builds=sbnd_ci/8454&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/14212&builds=

FNALbuild commented:
-icarus tests warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:icarus/view_builds/index?offset=0&builds=icarus_ci/3508&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/14212&builds=

FNALbuild commented:
-dune tests warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/12621&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/14212&builds=

FNALbuild commented:
-uboone tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/7012&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/14212&builds=

lgarren commented:
@therealendrju  the uboonecode CI tests are failing.  The problem may be in ubsim/ubsim/EventGenerator.

lgarren commented:
@therealendrju nevermind.  The CI test failures exist when this PR is not included.  Sorry for the noise.

absolution1 commented:
approve

FNALbuild commented:
This pull request is fully signed and it will be merged to develop and built in the next LArSoft release after it passes the integration tests. This pull request will now be reviewed by the release team before it's merged.)

lgarren commented:
approve
